### PR TITLE
Add ServiceAgreements to CivilServant

### LINKED ISSUE #174: Civil Servant sees not yet agreed service in (confirmed) services list

### PR DESCRIPTION
**Pull request by TheFehr**
Closes #174﻿

### CODE REVIEW

File: .semaphore/semaphore.yml [hunk 1]
```diff
@@ -60,16 +60,16 @@ blocks:
             - bin/lint
         - name: Unit Tests
           commands:
-            - sem-service start postgres
-            - 'bundle exec rails db:create db:schema:load'
+            - sem-service start postgres 12
```
Comment (lukasbischof) [inline .semaphore/semaphore.yml:63]: Why not 13?

File: app/models/service.rb [hunk 2]
```diff
@@ -14,8 +14,11 @@ class Service < ApplicationRecord
   scope :at_date, ->(date) { where(arel_table[:beginning].lteq(date)).where(arel_table[:ending].gteq(date)) }
   scope :at_year, ->(year) { overlapping_date_range(Date.new(year), Date.new(year).at_end_of_year) }
   scope :active, -> { where('beginning <= ?', Time.zone.now).where('ending >= ?', Time.zone.now) }
-  scope :agreement, -> { where(civil_servant_agreed: false).or(where(organization_agreed: false)) }
-  scope :definitive, -> { where(civil_servant_agreed: true).where(organization_agreed: true) }
+  scope :civil_servant_agreement_pending, -> { where(civil_servant_agreed: nil) }
+  scope :organization_agreement_pending, -> { where(organization_agreed: nil) }
+  scope :agreement, -> { civil_servant_agreement_pending.or(organization_agreement_pending) }
+  scope :declined, -> { where(organization_agreed: false).or(civil_servant_agreed: false) }
```
Comment (lukasbischof) [inline app/models/service.rb:20]: Does that scope make sense? Do you even use it?

File: app/models/service.rb [hunk 3]
```diff
@@ -14,8 +14,11 @@ class Service < ApplicationRecord
...
+  scope :definitive, -> { where(organization_agreed: true).where(civil_servant_agreed: true) }
```
Comment (lukasbischof) [inline app/models/service.rb:21]:
```suggestion
  scope :definitive, -> { where(organization_agreed: true, civil_servant_agreed: true) }
```

File: config/routes.rb [hunk 4]
```diff
@@ -32,6 +32,10 @@
     get '/', to: 'overview#index'
 
     resource :civil_servant, only: %i[edit update], on: :collection
+    resources :service_agreements, only: :index do
+      post 'accept', to: 'service_agreements#accept'
+      post 'decline', to: 'service_agreements#decline'
```
Comment (lukasbischof) [inline config/routes.rb:37]: Wouldn't that be a patch/put route?

File: spec/factories/services.rb [hunk 5]
```diff
@@ -10,19 +10,22 @@
...
     civil_servant_agreed { true }
-    civil_servant_agreed_on { '2018-08-05' }
     organization_agreed { true }
-    organization_agreed_on { '2018-07-05' }
     association :civil_servant, :full
 
+    trait :decided_dates_set do
+      civil_servant_decided_at { 2.weeks.ago }
+      organization_decided_at { 3.weeks.ago }
+    end
```
Comment (lukasbischof) [inline spec/factories/services.rb:19]: I don't really get why this trait is necessary. Seems like it's only there because of the model spec. If it is because of that, I'd prefer to put it in the built model there directly

File: spec/models/service_spec.rb [hunk 6]
```diff
@@ -142,14 +142,77 @@
       let(:other_beginning) { (service_range.begin - 2.months).at_beginning_of_week }
       let(:other_ending) { (service_range.begin - 1.month).at_end_of_week - 2.days }
 
-      before { create :service, civil_servant: civil_servant, beginning: other_beginning, ending: other_ending }
+      context 'when existing service it is a service agreement' do
```
Comment (lukasbischof) [inline spec/models/service_spec.rb:145]: huh, I do not get the context description 😄

File: spec/models/service_spec.rb [hunk 7]
```diff
@@ -142,14 +142,77 @@
...
+      end
+    end
+
+    describe '#legitimate_civil_servant_decision' do
+      context 'when the civil servant has not yet decided' do
+        subject do
+          service = build(:service, :civil_servant_agreement_pending)
+          service.update(civil_servant_agreed: true)
+          service.errors.added?(:civil_servant_agreed, :already_decided)
+        end
```
Comment (lukasbischof) [inline spec/models/service_spec.rb:183]:
```suggestion
        let(:service) { build(:service, :civil_servant_agreement_pending, civil_servant_agreed: true)

        subject do
          service.tap(&:validate).errors.added?(:civil_servant_agreed, :already_decided)
        end
```

File: spec/models/service_spec.rb [hunk 8]
```diff
@@ -142,14 +142,77 @@
...
+
+        it { is_expected.to be false }
+      end
+
+      context 'when the civil servant has already decided' do
+        subject do
+          service = build(:service, :decided_dates_set)
+          service.update(civil_servant_agreed: true)
+          service.errors.added?(:civil_servant_agreed, :already_decided)
+        end
```
Comment (lukasbischof) [inline spec/models/service_spec.rb:193]: here too, you can even reuse the let

File: spec/models/service_spec.rb [hunk 9]
```diff
@@ -197,20 +260,19 @@
     end
   end
 
-  describe 'active record hooks' do
+  describe 'database triggers' do
```
Comment (lukasbischof) [inline spec/models/service_spec.rb:263]: Wouldn't you also need to test the create trigger?

File: .semaphore/semaphore.yml (hunk 1 above)
Comment (TheFehr) [reply .semaphore/semaphore.yml:63]: Cause heroku uses 12

File: app/models/service.rb (hunk 2 above)
Comment (TheFehr) [reply app/models/service.rb:20]: Use? Nope.
Make sense? Not sure, but I'll remove I can always readd it in future if we need it

File: spec/factories/services.rb (hunk 5 above)
Comment (TheFehr) [reply spec/factories/services.rb:19]: True, I'll change it

File: spec/models/service_spec.rb (hunk 7 above)
Comment (TheFehr) [reply spec/models/service_spec.rb:183]: Ur suggesten forgoes the needed update

File: config/routes.rb (hunk 4 above)
Comment (TheFehr) [reply config/routes.rb]: But now my ui is broken

File: spec/models/service_spec.rb (hunk 7 above)
Comment (lukasbischof) [reply spec/models/service_spec.rb:183]: Doesn't the trigger also trigger on create?

File: app/controllers/civil_servants/service_agreements_controller.rb [hunk 10]
```diff
@@ -28,9 +28,18 @@ def decline
...
+        flash[:error] = format_errors(t('.erroneous_decline'), service)
       end
       redirect_to civil_servants_service_agreements_path
     end
+
+    private
+
+    # :reek:UtilityFunction
+    def format_errors(error_header, service)
+      return '' if service.valid?
```
Comment (lukasbischof) [inline app/controllers/civil_servants/service_agreements_controller.rb:40]: Why is this necessary? I'd guess that when it reached this method, the service is invalid anyways?

File: app/controllers/civil_servants/service_agreements_controller.rb [hunk 11]
```diff
@@ -28,9 +28,18 @@ def decline
...
+
+      error_header + service.errors.full_messages.join
```
Comment (lukasbischof) [inline app/controllers/civil_servants/service_agreements_controller.rb:42]:
```suggestion
      error_header + service.errors.full_messages.to_sentence
```

**Review by lukasbischof — APPROVED** [review]
me appoove
